sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -122,6 +122,37 @@ object JdbcUtils extends Logging {
     }
   }
 
+  /**
+   * Runs a custom query against a table from the JDBC database.
+   */
+  def runQuery(conn: Connection, actions: String, options: JDBCOptions): Unit = {
+    val autoCommit = conn.getAutoCommit
+    conn.setAutoCommit(false)
+    val queries = actions.split(";")
+    try {
+      queries.foreach { query =>
+        val queryString = query.trim()
+        val statement = conn.prepareStatement(queryString)
+        try {
+          statement.setQueryTimeout(options.queryTimeout)
+          val hasResultSet = statement.execute()
```

Comment:
@moomindani Is there a reason to allow a query here ? I am asking, since no one can consume this result set ? Or we are allowing it, just coz we are not able to parse the input to know what kind of statement it is ?

Reply:
It is to support stored procedures. (related to https://issues.apache.org/jira/browse/SPARK-32014)
For example, MySQL does not return result sets in stored procedures. In such case, we can use `executeUpdate()`.
However, some other databases returns result sets in stored procedures. To support such use-case, we need to consider both `executeQuery()` and `executeUpdate()`. That's the reason why I used `execute()` here.

Reply:
@moomindani I see. But if we call a cursory stored procedure in the pre/post action, what do we do with the result set retuned from it ? I am thinking users would call a cursory stored procedure if they want to do something with the results, no ? There are also some engines that can return multiple result set from a stored procedure.

https://stackoverflow.com/questions/53621106/db2-multiple-result-set-stored-procedure

Reply:
Currently result sets are always ignored if `preActions` or `postActions` return them. As you said, cursor won't be able to be used for reading as DataFrame.

By design, basically DDL queries are expected for `preActions` and `postActions`.
The reason why I used `execute` instead of `executeUpdate` is just to avoid exceptions due to stored procedure which returns result sets. Not to use result sets as an input for DataFrame.

Do you think it is enough for us to explain that result sets are always ignored in `preActions` and `postActions` if query results the result sets?
Or should we narrow down use-cases to support only DDL queries (which do not return result sets) by using `executeUpdate` instead of `execute`?

My idea was the first one, but I want to follow community's opinion.

Reply:
@moomindani 
IMHO, the best behavior would be to only allow DDL or DML (insert/update/delete) in pre and post actions. But if its not achievable, then we have to clearly document what is allowed and ramifications of users specifying other kind of statements. Lets please get input from @maropu and @HyukjinKwon to decide on the best course to move forward.

Reply:
I really appreciate your kind review. I had a quick conversation with @maropu and he agreed with your idea. 
After reviewing your comment, I changed my mind to allow only DDL and DML (insert/update/delete) queries in `preActions/postActions` to use `addBatch()` and `executeBatch()` instead of looping `execute()`. I reflected changes in the codes and also updated docs.

Reply:
@moomindani The changes look reasonable to me. One small question. Read the following :
```
JDBC drivers are not required to support this feature. You should use the 
DatabaseMetaData.supportsBatchUpdates() method to determine if the
target database supports batch update processing. The method returns 
true if your JDBC driver supports this feature.
```
at [link](https://www.tutorialspoint.com/jdbc/jdbc-batch-processing.htm)

So can we assume all the underlying JDBC implementations we have support this
feature ? As a start, we could just check DatabaseMetaData.supportsBatchUpdates() and 
raise an error saying pre and post actions are not supported ? Or we could address it as
a follow-up ?

Reply:
Your concern is valid, currently most of JDBC drivers (including MySQL, PostgreSQL, Oracle, SQL Server, Redshift, Athena, Impala, Snowflake, etc.) support batchUpdate, but I guess not all drivers support it.
My current implementation followed existing codes in `JdbcUtils.savePartition()` since it already used `addBatch()` and `executeBatch()`.

I could come up with following 3 possible directions for it.
(a) Keep current implementation. The `preActions`/`postActions` are only supported in JDBC drivers which support batchUpdates.
(b) Loop `executeUpdate()` instead of using `addBatch()`/`executeBatch()`. Although it sacrifices performance benefit, it will be safer when JDBC drivers do not support batchUpdates.
(c) Add check logic to use `DatabaseMetaData.supportsBatchUpdates()`. 
  - (c-1). Add it only to `preActions`/`postActions`
  - (c-2). For consistency, add it into both `preActions`/`postActions` and `JdbcUtils.savePartition()`.

Personally I prefer (a) or (b). 
If we do (c), we might need to implement two different ways to send queries.

Reply:
@moomindani Option `(a)` sounds okay to me as its simpler. We can always improve the behavior later. 
wdyt @maropu  ?

Reply:
Yea, `(a)` sounds fine to me.

> Only DDL or DML (insert/update/delete) are allowed. 

Btw, how do we control this? what if SELECT queries given?

Reply:
@maropu 
My expectation is that the JDBC driver would reject a statement other than DML (insert/update/delete). Looking at this link https://stackoverflow.com/questions/24577358/how-to-add-select-statements-in-addbatch, it seems like executeBatch may also reject DDLS.

@moomindani Can you please confirm if we can execute DDL's in executeBatch ?

Reply:
> Btw, how do we control this? what if SELECT queries given?

I confirmed that it returns exception like `Method is not allowed for a query`.

> Can you please confirm if we can execute DDL's in executeBatch ?

Sure, let me confirm it and add test to use DDL.

Reply:
I added some tests for DDLs. As far as I tested, `addBatch()`/`executeBatch()` worked also for DDLs.
It seems it also works for MySQL Connector/J since `executeBatch` just calls `executeUpdate` internally. https://github.com/mysql/mysql-connector-j/blob/dbd865c9fc68701efc70197ea1e65002520bed91/src/main/user-impl/java/com/mysql/cj/jdbc/StatementImpl.java#L859
I found a blog post which used `addBatch()`/`executeBatch()` with MySQL. https://www.roseindia.net/tutorial/java/jdbc/jdbcbatch/jdbcbatchdatabasetablecreation.html

However, I could not find any clear documentation which says DDLs are supported in `executeBatch()`. What I could find is that DDLs are supported in `executeUpdate()` and `executeLargeUpdate()`.
https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/Statement.html

Should we move from (a) to (b) as safer approach?